pkg/scheduler/plugins/elasticquota/plugin.go

```diff
@@ -0,0 +1,565 @@
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/listers/core/v1"
+	policylisters "k8s.io/client-go/listers/policy/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/api/v1/resource"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	"k8s.io/kubernetes/pkg/scheduler/util"
+	schedulerv1alpha1 "sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
+	"sigs.k8s.io/scheduler-plugins/pkg/generated/clientset/versioned"
+	"sigs.k8s.io/scheduler-plugins/pkg/generated/informers/externalversions"
+	"sigs.k8s.io/scheduler-plugins/pkg/generated/listers/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	"github.com/koordinator-sh/koordinator/apis/scheduling/config"
+	"github.com/koordinator-sh/koordinator/apis/scheduling/config/validation"
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/plugins/elasticquota/core"
+)
+
+const (
+	Name             = "ElasticQuota"
+	RevokePodCycle   = time.Duration(1) * time.Second
+	SnapshotStateKey = "ElasticQuotaSnapshot"
+)
+
+type Plugin struct {
+	sync.RWMutex
+	handle      framework.Handle
+	pluginArgs  *config.ElasticQuotaArgs
+	quotaLister v1alpha1.ElasticQuotaLister
+	podLister   v1.PodLister
+	pdbLister   policylisters.PodDisruptionBudgetLister
+	nodeLister  v1.NodeLister
+	//only used in OnNodeAdd,in case Recover and normal Watch double call OnNodeAdd
+	nodeResourceMapLock sync.Mutex
+	nodeResourceMap     map[string]struct{}
+	groupQuotaManager   *core.GroupQuotaManager
+	// quotaOverUsedRevokeController's add/erase pod only in Reserve/ Unreserve
+	quotaOverUsedRevokeController *QuotaOverUsedRevokeController
+	// elasticQuotaController only used to update elastic quota crd
+	elasticQuotaController *Controller
+	// PreFilter, refresh elasticQuotaInfos' runtime/used according to groupQuotaManager, then write snapshot.
+	// Reserve/ Unreserve/ OnPodAdd/ OnPodUpdate/ OnPodDelete, update the corresponding SimpleElasticQuotaInfo's pods.
+	elasticQuotaInfos SimpleElasticQuotaInfos
+	failOver          chan struct{}
+}
+
+// SnapshotState stores the snapshot of elasticQuotaInfos.
+type SnapshotState struct {
+	elasticQuotaInfos SimpleElasticQuotaInfos
+}
+
+// Clone the ElasticQuotaSnapshot state.
+func (s *SnapshotState) Clone() framework.StateData {
+	return &SnapshotState{
+		elasticQuotaInfos: s.elasticQuotaInfos.clone(),
+	}
+}
+
+var (
+	_ framework.PreFilterPlugin  = &Plugin{}
+	_ framework.PostFilterPlugin = &Plugin{}
+	_ framework.ReservePlugin    = &Plugin{}
+)
+
+func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	pluginArgs, ok := args.(*config.ElasticQuotaArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type GangSchedulingArgs, got %T", args)
+	}
+	if err := validation.ValidateElasticQuotaArgs(pluginArgs); err != nil {
+		return nil, err
+	}
+
+	client, ok := handle.(versioned.Interface)
+	if !ok {
+		client = versioned.NewForConfigOrDie(handle.KubeConfig())
```

Comment:
The contentType of handle.Kubeconfig will be configured as protobuf, which does not support CRD. You can modify it with reference to the implementation of other plugins.

Reply:
fixed